File: Cpp/Z Algorithm/z.cpp
//
// Created by Lazy Cat on 2024/4/3.
//
#include <vector>
#include <string>
using namespace std;

class Solution {
public:
    long long sumScores(string s) {
        int n = s.size();
        int left = 0, right = 0;
        vector<int> z(n, 0);
        long long ans = n;
        for(int i = 1; i < n; ++i){
            if(i <= right){
                z[i] = min(z[i - left], right - i + 1);
            }
            while(z[i] + i < n && s[z[i]] == s[z[i] + i]){
                left = i;
                right = i + z[i]++;
            }
            ans += z[i];
        }
        return ans;
    }
};

File: Py/Z Algorithm/z.py


class Solution:
    def sumScores(self, s: str) -> int:
        n = len(s)
        z = [0] * n
        ans = n
        left = right = 0
        for i in range(1, n):
            if i <= right:
                z[i] = min(right - i + 1, z[i - left])
            while i + z[i] < n and s[z[i]] == s[i + z[i]]:
                left, right = i, i + z[i]
                z[i] += 1
            ans += z[i]
        return ans